### PR TITLE
NODE-551 DEX watches balances and cancels invalid orders

### PR DESCRIPTION
**Pull request by vsuharnikov**

### CODE REVIEW

File: src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    portfolios
+      .get(ownerAddress)
+      .map { portfolio =>
+        val ordersByPriority = orders.sortBy(_.order.timestamp)(Ordering[Long].reverse)
+        val (_, r) = ordersByPriority.foldLeft((portfolio, List.empty: OrdersToDelete)) {
+          case ((restPortfolio, toDelete), limitOrder) =>
+            val updatedPortfolio = restPortfolio
+              .copy(balance = restPortfolio.balance - restPortfolio.leaseInfo.leaseOut)
+              .remove(limitOrder.spentAcc.assetId, limitOrder.getSpendAmount)
+              .flatMap(_.remove(None, limitOrder.remainingFee))
```
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala]: I guess orders which buy Waves for BTC wouldn't work here if there are not enough waves for a fee.

File: src/main/scala/com/wavesplatform/utx/MatcherUtxPool.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import scala.collection.mutable
+
+class MatcherUtxPool(underlying: UtxPool, events: EventStream) extends UtxPool with ScorexLogging {
+  override def putIfNew(tx: Transaction): Either[ValidationError, Boolean] = {
+    underlying
+      .putIfNew(tx)
+      .map { r =>
+        if (r) {
+          tx match {
```
Comment (koloale) [inline src/main/scala/com/wavesplatform/utx/MatcherUtxPool.scala]: Maybe exclude exchange transactions from reacting

File: src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    context.system.scheduler.scheduleOnce(TimeoutToProcessChanges, self, ReceiveTimeout)
+  }
+
+  private def replaceChanges(orig: ChangesByAddress, replacement: ChangesByAddress): ChangesByAddress = orig ++ replacement
+
+  private def ordersToDelete(portfolios: Map[String, Portfolio], ownerAddress: String, orders: Seq[LimitOrder]): OrdersToDelete = {
+    portfolios
+      .get(ownerAddress)
+      .map { portfolio =>
+        val ordersByPriority = orders.sortBy(_.order.timestamp)(Ordering[Long].reverse)
```
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala]: No need to check all orders, only with spentAsset in received changes.

File: src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  private def waitOrders(taskId: Long,
+                         stashedChanges: ChangesByAddress,
+                         changesByAddress: ChangesByAddress,
+                         waitOrdersTimeout: Cancellable): Receive = {
+    case GetActiveOrdersByAddressResponse(`taskId`, address, orders) =>
+      ordersToDelete(changesByAddress, address, orders)
+        .foreach {
+          case (pair, id) =>
+            orderHistory ! ForceCancelOrderFromHistory(id)
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala]: Sad, but it doesn't work here. If we didn't send this message to the order history, we get the failed tests "owner moves assets/waves to another account and order become an invalid" in MatcherTestSuite. The state of the order in history isn't changed:
```
2018-03-02 10:24:49.589 DEBUG c.w.i.D.DockerNode.node01 - Awaiting condition 'Order 'BNn1mR5Miqar6j848yUNPnPP8caWe6RyLmeQYiwPaUK6' is cancelled'
2018-03-02 10:24:49.589 TRACE c.w.i.D.DockerNode.node01 - Executing request 'http://localhost:33901/matcher/orderbook/NewBobCoin/WAVES/BNn1mR5Miqar6j848yUNPnPP8caWe6RyLmeQYiwPaUK6	GET	headers:'
2018-03-02 10:24:49.622 DEBUG c.w.i.D.DockerNode.node01 - Request: http://localhost:33901/matcher/orderbook/NewBobCoin/WAVES/BNn1mR5Miqar6j848yUNPnPP8caWe6RyLmeQYiwPaUK6
Response: {
  "status" : "Accepted"
}
...
2018-03-02 10:25:48.976 DEBUG c.w.i.D.DockerNode.node01 - Request: http://localhost:33901/matcher/orderbook/NewBobCoin/WAVES/BNn1mR5Miqar6j848yUNPnPP8caWe6RyLmeQYiwPaUK6
Response: {
  "status" : "Accepted"
}
```

File: src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala (hunk 1 above)
Comment (vsuharnikov) [reply src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala]: 👍 I'll create a test and fix this

**Pull request merged**